internal/storage/serde_events.go

```diff
@@ -0,0 +1,575 @@
+	closers []func()
+	fields  []FieldID
+
+	r compositeRecord
+}
+
+func (crr *compositeBinlogRecordReader) iterateNextBatch() error {
+	if crr.closers != nil {
+		for _, close := range crr.closers {
+			if close != nil {
+				close()
+			}
+		}
+	}
+	crr.blobPos++
+	if crr.blobPos >= len(crr.blobs[0]) {
+		return io.EOF
+	}
+
+	for i, b := range crr.blobs {
+		reader, err := NewBinlogReader(b[crr.blobPos].Value)
+		if err != nil {
+			return err
+		}
+
+		crr.fields[i] = reader.FieldID
+		// TODO: assert schema being the same in every blobs
+		crr.r.schema[reader.FieldID] = reader.PayloadDataType
+		er, err := reader.NextEventReader()
+		if err != nil {
+			return err
+		}
+		rr, err := er.GetArrowRecordReader()
+		if err != nil {
+			return err
+		}
+		crr.rrs[i] = rr
+		crr.closers[i] = func() {
+			rr.Release()
+			er.Close()
+			reader.Close()
+		}
+	}
+	return nil
+}
+
+func (crr *compositeBinlogRecordReader) Next() error {
+	if crr.rrs == nil {
+		if crr.blobs == nil || len(crr.blobs) == 0 {
+			return io.EOF
+		}
+		crr.rrs = make([]array.RecordReader, len(crr.blobs))
+		crr.closers = make([]func(), len(crr.blobs))
+		crr.blobPos = -1
+		crr.fields = make([]FieldID, len(crr.rrs))
+		crr.r = compositeRecord{
+			recs:   make(map[FieldID]arrow.Record, len(crr.rrs)),
+			schema: make(map[FieldID]schemapb.DataType, len(crr.rrs)),
+		}
+		if err := crr.iterateNextBatch(); err != nil {
+			return err
+		}
+	}
+
+	composeRecord := func() bool {
+		for i, rr := range crr.rrs {
+			if ok := rr.Next(); !ok {
+				return false
+			}
+			// compose record
+			crr.r.recs[crr.fields[i]] = rr.Record()
+		}
+		return true
+	}
+
+	// Try compose records
+	if ok := composeRecord(); !ok {
+		// If failed the first time, try iterate next batch (blob), the error may be io.EOF
+		if err := crr.iterateNextBatch(); err != nil {
+			return err
+		}
+		// If iterate next batch success, try compose again
+		if ok := composeRecord(); !ok {
+			// If the next blob is empty, return io.EOF (it's rare).
+			return io.EOF
+		}
+	}
+	return nil
+}
+
+func (crr *compositeBinlogRecordReader) Record() Record {
+	return &crr.r
+}
+
+func (crr *compositeBinlogRecordReader) Close() {
+	for _, close := range crr.closers {
+		if close != nil {
+			close()
+		}
+	}
+}
+
+func parseBlobKey(blobKey string) (colId FieldID, logId UniqueID) {
+	if blobKey == "" {
+		return 0, 0
+	}
+	if _, _, _, colId, logId, ok := metautil.ParseInsertLogPath(blobKey); ok {
+		return colId, logId
+	}
+	if colId, err := strconv.ParseInt(blobKey, 10, 64); err == nil {
+		// data_codec.go generate single field id as blob key.
+		return colId, 0
+	}
+	return -1, -1
+}
+
+func newCompositeBinlogRecordReader(blobs []*Blob) (*compositeBinlogRecordReader, error) {
+	sort.Slice(blobs, func(i, j int) bool {
+		iCol, iLog := parseBlobKey(blobs[i].Key)
+		jCol, jLog := parseBlobKey(blobs[j].Key)
+
+		if iCol == jCol {
+			return iLog < jLog
+		}
+		return iCol < jCol
+	})
+
+	blobm := make([][]*Blob, 0)
```

Comment:
this uses an undefined slice size, which might need the scaled capacity to store and cause copy operations while keeping appending. Maybe it is better to use the allocated map. 
```
func newCompositeBinlogRecordReader(blobs []*Blob) (*compositeRecordReader, error) {
	blobMap := make(map[FieldID][]*Blob)
	for _, blob := range blobs {
		colId, _ := parseBlobKey(blob.Key)
		if _, exists := blobMap[colId]; !exists {
			blobMap[colId] = []*Blob{blob}
		} else {
			blobMap[colId] = append(blobMap[colId], blob)
		}
	}
	sortedBlobs := make([][]*Blob, 0, len(blobMap))
	for _, blobsForField := range blobMap {
		sort.Slice(blobsForField, func(i, j int) bool {
			_, iLog := parseBlobKey(blobsForField[i].Key)
			_, jLog := parseBlobKey(blobsForField[j].Key)

			return iLog < jLog
		})
		sortedBlobs = append(sortedBlobs, blobsForField)
	}
	return &compositeRecordReader{
		blobs: sortedBlobs,
	}, nil
}
```